### ISSUE TITLE
Bug: doesNotContainTypeClasses is missing detection of some type classes after #40

### ISSUE THREAD

**Issue by harendra-kumar** [nomeata/inspection-testing#41]
PR #40 seems to have introduced a bug because of which some typeclasses are not detected/reported even if they are not in the exception list.

It works fine in the older version of the PR that used `ExceptT` but the newer list/concatMap based version has the bug. On a quick look I have not been able to figure out what the issue might be here.

@nomeata can you take a look if you can find anything.

I have this test:

```
inspect $ hasNoTypeClassesExcept 'fromStreamVar
    [ ''Monad
    , ''Applicative
    , ''MonadThrow
    , ''Exception
    , ''MonadIO
    ]
```

The old code fails because it finds MonadBaseControl type class as well as Typeable and Functor in the core. However, the new code is unable to find these and passes happily.

You can find the code [here](https://github.com/composewell/streamly/blob/55bdd5b8d57911e2bca7130ad4f8a8aaa22018e2/src/Streamly/Internal/Data/Stream/SVar.hs#L110). Just remove the last typeclass in that list and build the repo with `cabal build --flag inspection`, the new code succeeds whereas the old code fails.

The following ExceptT version works as expected though:

```haskell
allTyCons :: (TyCon -> Bool) -> Slice -> Maybe (Var, CoreExpr, TyCon)
allTyCons predicate slice =
    listToMaybe $ do
        (v,e) <- slice
        case runExcept (go e) of
            Right _ -> A.empty
            Left tc -> return (v,e,tc)
  where

    success = return undefined
    failure tc = throwE tc

    runAll f = foldr (\x xs -> f x >> xs) success

    goV v = goT (varType v)

    go (Var v)           = goV v
    go (Lit _)           = success
    go (App e a)         = go e >> go a
    go (Lam b e)         = goV b >> go e
    go (Let bind body)   = runAll goB (flattenBinds [bind]) >> go body
    go (Case s b _ alts) = go s >> goV b >> runAll goA alts
    go (Cast e _)        = go e
    go (Tick _ e)        = go e
    go (Type t)          = (goT t)
    go (Coercion _)      = success

    goB (b, e) = goV b >> go e

    goA (_,pats, e) = runAll goV pats >> go e

    goT (TyVarTy _)      = success
    goT (AppTy t1 t2)    = goT t1 >> goT t2
    goT (TyConApp tc ts) = if predicate tc then runAll goT ts else failure tc
                           -- ↑ This is the crucial bit
    goT (ForAllTy _ t)   = goT t
#if MIN_VERSION_GLASGOW_HASKELL(8,2,0,0)
    goT (FunTy
# if MIN_VERSION_GLASGOW_HASKELL(8,9,0,0)
               _
# endif
                 t1 t2)  = goT t1 >> goT t2
#endif
    goT (LitTy _)        = success
    goT (CastTy t _)     = goT t
    goT (CoercionTy _)   = success
```
